site/source/docs/debugging/Sanitizers.rst

```diff
@@ -378,3 +372,56 @@ option ``malloc_context_size=0``, like this:
 This prevents ASan from reporting the location of memory leaks, as well as
 offer insight into where the memory for a heap-based memory error originated,
 but may provide tremendous speed ups.
+
+Comparison to ``SAFE_HEAP``
+---------------------------
+
+Emscripten provides a ``SAFE_HEAP`` mode, which can be activated by running
+``emcc`` with ``-s SAFE_HEAP``. This does several things, some of which overlap
+with sanitizers.
+
+In general, ``SAFE_HEAP`` focuses on the specific pain points that come up when
+targeting wasm. The sanitizers on the other hand focus on the specific pain points that are
+involved with using languages like C/C++. Those two sets overlap, but are not
+identical. Which you should use depends on which types of problems you are
+looking for. You may want to test with all sanitizers and with ``SAFE_HEAP``
+for maximal coverage, but you may need to build separately for each mode, since
+not all sanitizers are compatible with each other, and not all of them are
+compatible with ``SAFE_HEAP`` (because the sanitizers do some pretty radical
+things!). You will get a compiler error if there is an issue with the flags you
+passed. A reasonable set of separate test builds to do might be: ASan, UBsan,
+and ``SAFE_HEAP``.
+
+The specific things ``SAFE_HEAP`` errors on include:
+
+* **NULL pointer (address 0) reads or writes**. As mentioned earlier, this is
+  annoying in WebAssembly and JavaScript because 0 is just a normal address, so
+  you don't get an immediate segfault, which can be confusing.
+* **Unaligned reads or writes**. These work in WebAssembly, but on some platforms
+  an incorrectly-aligned read or write may be much slower, and with wasm2js
+  (``WASM=0``) it will be incorrect, as JavaScript Typed Arrays do not allow
+  unaligned operations.
+* **Reads or writes past the top of valid memory** as managed by ``sbrk()``, that is,
+  memory that was not properly allocated by ``malloc()``. This is not specific
+  to wasm, however, in JavaScript if the address is big enough to be outside the
+  Typed Array, ``undefined`` is returned which can be very confusing, which is
+  why this was added (in wasm at least an error is thrown; ``SAFE_HEAP`` still
+  helps with wasm though, by checking the area between the top of ``sbrk()``'s
+  memory and the end of the wasm Memory).
+
+``SAFE_HEAP`` does these checks by instrumenting every single load and store.
+That has the cost of slowing things down, but it does give a simple guarantee
+of finding *all* such problems. It can also be done after compilation, on an
+arbitrary wasm binary, while the sanitizers must be done when compiling from
+source.
+
+In comparison, UBSan can also find null pointer reads and writes. It does not
+instrument every single load and store, however, as it is done during
+compilation of the source code, so the checks are added where clang knows they
+are needed. This is much more efficient, but there is a risk of codegen and
```

Comment:
Curious - is this a real risk / does it happen in the wild? If it does, is this usually considered a bug in Clang?

Reply:
It could be either a literal bug, or undefined behavior that is taken advantage of (probably more likely). Either way it could surprise users I think.

Reply:
Huh, but isn't the point of UBSan to find that undefined behavior before it's taken advantage of? (I admit I don't know specifics / difference, just judging by naming)

Reply:
I'm not totally sure either about this. That's a good point and it could be the case.